### ISSUE TITLE
Workspace: Element Keyboard Navigation and Panel Focus

### ISSUE THREAD

**Issue by carloskelly13** [GoogleForCreators/web-stories-wp#5619]
## Feature Description

Currently there is logic to prevent an element on the canvas from moving when the arrow keys are pressed. Now that panels are evolving in functionality we likely need to expand this to other components and controls outside of the canvas as well.

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

**Comment by miina** [comment]
It looks like in case of any inputs (and also the dropdowns that are working as expected now but were not when the issue was initially created), the element does not move when using the arrow keys.
It does move, however, when focusing on any toggles/buttons, e.g. after clicking on text alignment button and then using the arrow keys, the element does move. But that seems to make sense in terms of accessibility. Thoughts, @barklund / @swissspidy (let me know if you think otherwise)?

**Comment by swissspidy** [comment]
That seems reasonable at first glance, yeah. Not sure I 100% understand the original ticket description though. Do we need to check for this in other places beyond the design panel?

**Comment by barklund** [comment]
When the keyboard focus is somewhere in the design/document panel, we generally try to stick to the rule that either the focused element handle all arrow key inputs or none (in which case they're passed to the selected element on the canvas if any). This was a decision made long ago, but it's definitely up for discussion.

If we stick by this decision, we still have a few mistakes though, that have crept in over time. Going through various elements and trying to use the arrow keys while focused, I discovered a few weird instances:

1) The new dropdowns cannot be opened using the up  or down arrow, which is normally a common way to open a closed but focused dropdown (and [WAI-ARIA recommended](https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-collapsible.html#:~:text=the%20next%20option.-,If%20the%20listbox%20is%20collapsed%2C%20also%20expands%20the%20list.,-Up%20Arrow)). Instead, the active element moves:
![Screen Shot 2021-11-05 at 14 59 18](https://user-images.githubusercontent.com/637548/140564418-76ca7d98-b86c-4466-b17b-f4d1827e5b86.png)

2) The panel headers (basically accordion buttons) could optionally expand-collapse when interacted with using arrow keys. It's completely optional though, and actually not mentioned in the [WAI-ARIA recommendations for accordions](https://www.w3.org/TR/wai-aria-practices-1.1/#accordion) (but we also don't follow the recommendations strictly anyway). Alternatively, we could follow recommendations more strictly and then make arrow keys move between panels when on a panel header:
![Screen Shot 2021-11-05 at 14 59 31](https://user-images.githubusercontent.com/637548/140564430-bd2d014b-a3e0-4738-b2e0-23653ee77015.png)

3) The new taxonomy category selector in the document panel captures up-down arrow keys, but not left-right arrow keys, so pressing left or right while focusing this element moves the active element on the canvas, while pressing up or down changes focus inside the panel control:
![Screen Shot 2021-11-05 at 14 59 04](https://user-images.githubusercontent.com/637548/140564553-3327326b-ffa2-4ffa-b7c1-29c982146ea0.png)

4) The direction selector for animation (e.g. fly-in) captures all arrow keys but doesn't do anything with them. If keyboard focus is on one of the direction buttons, arrow keys are ignored. They should either do something in the panel control (change the direction) or be passed to the selected element:
![Screen Shot 2021-11-05 at 14 58 54](https://user-images.githubusercontent.com/637548/140564575-06ea3c02-cbf9-4798-8183-a163d5fe6580.png)

---
However, a completely different approach (following the above and/or instead of the above) could be to disable keyboard events from bubbling at all while focus is anywhere inside the design/document panel. It _is_ a little weird that you can have the keyboard focus on a random button over there and the active element moves around.

But on the other hand, if you use the mouse to click a button, the button will keep focus afterward, which would then disable keyboard movement of the active element which might also be unwanted. Should we unfocus buttons again if activated via mouse perhaps? Or is that a complete no-no?

There are many potential issues and ditto fixes here.

**Comment by miina** [comment]
> But on the other hand, if you use the mouse to click a button, the button will keep focus afterward, which would then disable keyboard movement of the active element which might also be unwanted. Should we unfocus buttons again if activated via mouse perhaps? Or is that a complete no-no?

This seems quite unexpected as well, the button should get the focus when clicking with mouse.

> However, a completely different approach (following the above and/or instead of the above) could be to disable keyboard events from bubbling at all while focus is anywhere inside the design/document panel. It is a little weird that you can have the keyboard focus on a random button over there and the active element moves around.

I wonder what's the general behaviour in these cases in editors, any thoughts? Perhaps @gajasinski / @agingoldseco might have insights here as well about if we should allow moving an element with the arrow keys at all if focus is anywhere inside the design/document panel.

**Comment by gajasinski** [comment]
Would need to take a look and dig in with @agingoldseco but my gut says to stick closely to the WAI-ARIA recommendations. Can find time to sit down and dig through these.
